### ISSUE TITLE
Incorrect keyboard layout

### ISSUE THREAD

**Issue by soleera** [google/mozc#427]
When using kana input in mozc-2.20.2673.102 with fcitx, mozc assumes a standard Japanese keyboard layout, even though the selected keyboard layout is US. this means that instead of 「 & 」being mapped to shift-゛ & shift-゜ as in MS-IME, they are instead mapped to shift-゜ & shift-む. The latter corresponds to shift-[ and shift-] on a Japanese keyboard, but on a US keyboard this corresponds to shift-] & shift-\ which is very confusing.

**Comment by Rybec** [comment]
Same problem here.  This is a pretty serious problem.  How is it that this was filed three years ago, and it doesn't even have a _reply?_

**Comment by Rybec** [comment]
Near the end of the following link, there are instructions for making Mozc use a Japanese keyboard layout.  This did _not_ work for me, but supposedly it did for other people.  I'll try a few more things, to see if I can find a more reliable work around, since clearly Google doesn't care.

The real problem here, is that Mozc doesn't have an option in the GUI config for setting the keyboard layout.  The actual config file seems to have a place where it can be set though.  Of course, like I said, this isn't working for me.

I'll report back, if I find anything.  Maybe that way the next person who finds this won't have to dig so much to find a solution.

**Comment by Rybec** [comment]
Ok, so I have made some progress, but it isn't ideal.  If you can manually set the keyboard to Japanese, through whatever external means your distro/desktop provides, you will get the right layout, with _one_ critical exception.  The `/~ key won't type ろ.  Instead, it will toggle between kana and English.  It is possible to disable this behavior, with a custom keymap, however, it still won't produce the desired behavior.*  In addition, if you use a custom keymap, you will need to make sure you bind "backspace" to the correct operation, as it won't work as expected otherwise.  This is a _really_ rough hack though.  It requires remembering to change your keyboard layout ever time you change input types.

* The ろ key on an actual Japanese keyboard is between "め" (/ and ? on US keyboards) and right shift.  No such key exists on U.S. keyboards, so ろ has to go somewhere else.  Switching to Japanese keyboard (since it expects a hardware keyboard) seems to just drop the key entirely.  On the other side, the `/~ key is replaced with a width/kanji toggle key (Japanese Apple keyboards are missing the key entirely).  So MS IME does the obvious thing, binding ろ to the `/~ key on U.S. keyboards.  Mozc does not bother, making it impossible to type correct Japanese on a U.S. keyboard, even _with_ the OS set to Japanese keyboard.  (And let me note, this means the MS-IME keybinding is _wrong,_ because actual MS-IME binds ろ to `/~ on U.S. keyboards.)

I might have a solution, but I have to figure out which Japanese key Mozc thinks `/~ is.  It _looks_ like it is being used for the "eisu" (英数) key, which is a special function of the Caps Lock key.  (I have no idea how this works on a real Japanese keyboard).  (Apple keyboards put it between left Command and the Spacebar.)  Anyhow, a custom keybinding or a modified version of the broken MS-IME keybindings, where the "eisu" functionality is mapped as an alternate function of Caps Lock (where it belongs), and the `/~ key is mapped to output ろ should fix this problem.  The only issue is, I cannot find any way of making a keybinding that will output an arbitrary character.  In short, as far as I can tell, Mozc is completely useless unless you have a legitimate Japanese keyboard.

Thanks Google.  I doubt you'll ever read this, and given your track record on this issue, I doubt even more you will bother fixing it in a timely manner, but you should at least be aware: Microsoft is actually _beating you_ on this one.

**Comment by Rybec** [comment]
For reference, here is what an actual (non Apple) Japanese keyboard looks like:

[Japanese Keyboard](https://en.wikipedia.org/wiki/Language_input_keys#/media/File:KB_Japanese.svg)


And here is how MS-IME maps Japanese to a U.S. keyboard:

[U.S. Keyboard mapping to Japanese, by the real MS-IME](https://filestore.community.support.microsoft.com/api/images/308fb616-d7aa-430f-96ce-761eb5617bdc?upload=true)

(In case that link to MS Support has issues, [here's the forum post it came from](https://answers.microsoft.com/en-us/windows/forum/all/japanese-ime-not-converting-properly/0780efac-8f1e-4dae-a1b2-079593f29e23).  You will find the image about halfway down.)

**Comment by Rybec** [comment]
I suppose it may be worth pointing out: [There are many alternative Japanese kayboard layouts.](http://xahlee.info/kbd/Japan_keyboard_layouts.html)

Any half decent Japanese input system should _probably_ be sufficiently customizable to allow the user to _completely redefine_ the entire keyboard layout, to fit whichever keyboard is being used.

**Comment by hiroyuki-komatsu** [comment]
Thank you for the reminder.

I had some investigations. IIUC, it seems there are two points.

- ibus-mozc always uses Kana layout for JP keyboard.
- Kana layout for US keyboard is not the same as Windows.

### ibus-mozc always uses Kana layout for JP keyboard.

If I understand correctly, the value of `layout_is_jp` is always true.
https://github.com/google/mozc/blob/master/src/unix/ibus/mozc_engine.cc#L353

Would you confirm whether the layout is changed when you modified the value of `name` in /usr/share/ibus/component/mozc.xml from 'mozc-jp' to others (e.g. `mozc-us`)?


### Kana layout for US keyboard is not the same as Windows.

This is because there is no standard of Kana keymap for US layout. The JIS standard only considers the usage with JP keyboard. As a result, Windows has a different Kana layout from ChromeOS or Mac. Mozc uses the same Kana layout with ChromeOS and Mac.

If you want to use the Kana input with a US keyboard as Windows style, I'd suggest trying either of the following ways.

- Customize romaji input table for Kana input, or
- Modify the source code


#### Customize romaji input table for Kana input

Open "Mozc Romaji table editor" as follows:
[tools] > [properties] > (open Mozc Settings) > [General] > [Romaji table] > [customize...]

Then, you can create Windows style Kana input with US keyboard.

mac-kana.tsv could be a good reference.
https://github.com/google/mozc/blob/master/src/data/preedit/mac-kana.tsv


#### Modify the source code

For ibus, the code is here. I think fcitx also has a similar code. 
https://github.com/google/mozc/blob/master/src/unix/ibus/key_translator.cc

Thanks,

**Comment by Rybec** [comment]
First, thank you for responding!  I actually didn't expect a response, but I wanted to provide what I could just in case.

Second, I am in Windows right now, but I will reboot into Linux tomorrow and check that.

Third, I understand that there isn't an official kana layout for U.S. keyboards.  What concerns me is that on U.S. keyboards, the layout assumed by mozc doesn't even have all of the kana typeable.  Obviously, I would prefer the layout Windows uses, because that's where I put the kana stickers, but anything that allows all kana to be typed would be better than the current situation.  (Ideally, there would be an option to choose the Windows or Apple layout, but again, anything that allows the full range of kana would be an improvement.)

Fourth, modifying the romaji table might be a good work around.  I hadn't thought of that.  It would still be a cheap hack that wouldn't be a universal fix.  (Can the romaji table be saved and loaded?  If so, then maybe a modified romaji table could be provided as a download.  If I end up going in that direction, I could probably provide that, if it is viable.)  Modifying the source code, is of course, always a good option.  It's not viable for me at this time though, due to lack of time.


I'll see if I can find some time tomorrow to work on this.  If so, I'll report back, and if I can find a way of saving the romaji table, I'll see if I can provide a copy, if I end up taking that route.

**Comment by shitamo** [comment]
Can save and load the romaji table.
![mozc-1](https://user-images.githubusercontent.com/62193758/99900913-2b011400-2cf6-11eb-8793-dec0f3b2073d.png)
![mozc-3](https://user-images.githubusercontent.com/62193758/99900918-33f1e580-2cf6-11eb-9790-454b10acf7c4.png)

**Comment by Rybec** [comment]
> Would you confirm whether the layout is changed when you modified the value of name in /usr/share/ibus/component/mozc.xml from 'mozc-jp' to others (e.g. mozc-us)?

I tried this, and it appears Ibus/Mozc just failed entirely.  The hotkey for changing input languages just plain doesn't work when I replace the default name with mozc-us.  Ibus config does not even show mozc as an installed input method.


@shitamo Thanks for answering that question.  I am only learning Japanese right now, so I can't read most of that, but it should be enough to get me to the right place.  (I am installing Japanese input on all of my devices that support it, to aid in learning.  I already use it a fair amount on my Android tablet, but I want to be able to use it in Linux on my laptop, since I tend to spend more time there.  Eventually, I hope to become proficient enough to change my display language to Japanese!)

Anyhow, I'll see if I can create a custom romaji mapping that matches the Windows IME layout for Japanese.  If I am successful, I'll export it and post it or a link to it here.  Due to the issue with there being no standard Japanese kana mapping to US keyboards, maybe providing alternate romaji mappings would be the best solution.

**Comment by Rybec** [comment]
This _almost_ works.  The following mapping can do all of the standard kana, without any issues.  It can also do all the punctuation shown on the MS-IME keyboard mapping.  Where it fails is small い, mapped to E (shift-e), small つ, mapped to Z (shift-z), and all dakuten and handakuten.

You will note in this mapping that most of the dakuten/handakuten are mapped to the base kana followed by [ or ] (the MS-IME mapping for the ten-ten and maru characters).  This didn't work, which is why I tried mapping the English key for か (t) and then [ to get が.  Typing t[ _does_ work correctly, however t by itself no longer prints か, without hitting the spacebar or following it with another character.  All of the other kana immediately display the appropriate kana when typed, but multi-character kana display the English characters typed, until all ambiguity is removed, making it really hard to follow my typing.  There is _almost_ a valid solution to this in Unicode.  U+3099 and U+309A are the combining ten-ten and maru, respectively, and mapping them to [ and ] _does work_, except that the produced characters are not the same as the single character versions.  For example, if I search for ず (the single unicode character for zu) in my editor (or my browser, Firefox), it won't match ず (す followed by combining ten-ten).  Yes, they _look_ identical, but the first is a unicode U+305A, while the second is a unicode U+3059 followed by U+3099.  If all I wanted to do was very basic communication, this would be fine, but a simple web search would fail to work correctly, and trying to find specific text within a document would be extremely problematic.

Is there maybe some special code I can use in the romaji mapping that will convert the previous kana to the dakuten or handakuten, if it has one?  Clearly I can't use nested expressions (using a kana produced by Mozc combined with another typed character, as I tried to do with the dakuten/handakuten in the provided mapping).  Maybe there is a way to do this in the keymap style?  (Though, it would be much more distributable if the mapping was confined to just one file...)


The shift case is unique, because it only seems to negatively affect _letter keys_.  The only kana that require the use of shift with a letter key are small い and small つ.  The only way I have found to type them with this layout is to type the character (which displays the capital English character initially) and then search through the list of suggestions to find the right character, which is extremely inconvenient, especially since they don't tend to appear anywhere near the top.  Shift with number keys for the rest of the small vowels and the small y- kana work perfectly fine, as does shift-0 for を and shift-- for ー, and the symbol keys work as expected with shift as well.  I have no idea how this issue might be fixed.  It seems really odd, in fact, that letter keys are treated so differently from pretty much everything else.

It may be worth noting that once the file has been imported, when I press OK to apply the new mapping, I get this message: 'Input fields contain capital letters.  "Shift-mode-switch" function is disabled with this new mapping.'  I assume this means that "shift" has some alternative function that is disabled to accommodate the use of capital letters.  The problem is, despite disabling this feature, it still fails to properly map the capital letters.  Unfortunately, small つ seems to be used fairly heavily in Japanese.  I am not sure about small い, as it is only used in katakana (in my limited experience), but Japanese has a lot of loan words that are likely to use it.


Anyhow, here is my mapping:

`
~	〜
+.	…
+,	‥
+h	←
+j	↓
+k	↑
+l	→
+[	『
+]	』
[	゙
]	゚と

t[	が
き[	ぎ
く[	ぐ
け[	げ
こ[	ご
さ[	ざ
し[	じ
す[	ず
せ[	ぜ
そ[	ぞ
た[	だ
ち[	ぢ
つ[	づ
て[	で
と[	ど
は[	ば
ひ[	び
ふ[	ぶ
へ[	べ
ほ[	ぼ
は]	ぱ
ひ]	ぴ
ふ]	ぷ
へ]	ぺ
ほ]	ぽ

`	ろ
1	ぬ
2	ふ
3	あ
#	ぁ
4	う
$	ぅ
5	え
%	ぇ
6	お
^	ぉ
7	や
&	ゃ
8	ゆ
*	ゅ
9	よ
(	ょ
0	わ
)	を
-	ほ
_	ー
=	へ
q	た
w	て
e	い
E	ぃ
r	す
t	か
y	ん
u	な
i	に
o	ら
p	せ
{	「
}	」
\	む
a	ち
s	と
d	し
f	はと
g	き
h	く
j	ま
k	の
l	り
;	れ
'	け
z	つ
Z	っ
x	さ
c	そ
v	ひ
b	こ
n	み
m	も
,	ね
<	、
.	る
>	。
/	め
?	・
`

**Comment by Rybec** [comment]
It is also worth noting:

When composing dakuten using the combining ten-ten, I get this for jitensha (bicycle): じてんしゃ

 (I seem to have broken something with Mozc, making it so I cannot change mappings, and I need to reboot.  I'll edit this to complete it once that is done.)

**Comment by Rybec** [comment]
On a side note, what exactly does the "next input" field do?  Is it possible there could be some solution using that?

**Comment by hiroyuki-komatsu** [comment]
> > Would you confirm whether the layout is changed when you modified the value of name in /usr/share/ibus/component/mozc.xml from 'mozc-jp' to others (e.g. mozc-us)?
>
> 
> I tried this, and it appears Ibus/Mozc just failed entirely. The hotkey for changing input languages just plain doesn't work when I replace the default name with mozc-us. Ibus config does not even show mozc as an installed input method.

Thank you for checking. I will try to make time to investigate it more.


### Shift-mode-switch

It's a customization field of Mozc settings located in [Advanced] > [Shift key mode switch]. It should be [Off], when the romaji table contains upper letters.

The current message is confusing. It'd be better to change it 
* from: 'Input fields contain capital letters. "Shift-mode-switch" function is disabled with this new mapping.'
* to: 'Input fields contain capital letters. "Shift-mode-switch" function should be [Off] with this new mapping.'

### next input

The characters of `next input` are prepended to your input characters.

Assuming your input is "itte" and the following table is used,
```
i <tab> い
tt <tab> っ <tab> t
te <tab> て
```

1. "i" of "itte" is converted to "い", and "tte" remains.
2. "tt" of "tte" is converted to "っ", and "t" (as the next input) is prepend to "e". So "te" remains.
3. "te" is converted to "て".
4. as a result, "itte" is converted to "いって"

So, to handle ten-ten, something like following would work:
```
t <tab><tab> か
か[ <tab> が
```

**Comment by hiroyuki-komatsu** [comment]
The commit 9ba59b64d53365c1fe93c1c245b4ec3e35bdadf0 fixes the issue that ibus-mozc always uses Kana layout for JP keyboard.

ibus_config.textproto will be created in ~/.mozc (or ~/.config/mozc), then you edit it to specify the keyboard layout.
If the layout is changed from `default` to `us`, Kana layout for US keyboard is used.

**Comment by hiroyuki-komatsu** [comment]
Closing this issue as 9ba59b64d53365c1fe93c1c245b4ec3e35bdadf0 provides a way to configure a keyboard layout without a root permission, and fixes an issue of Kana-input for US keyboard.

Additional comments and feedbacks are welcome.

**Comment by Rybec** [comment]
Thanks for the updates!  I'll take a look when I have time and report back if I find any issues.

**Comment by corigne** [comment]
Confirming, for US keyboards the layout setting of "us"  in the settings file _ibus_config.textproto_ fixed issues with mismatching keys compared to keycaps with printed kana sub legends, in case anyone comes around here wondering. Also thanks for documenting this.
